srcts/src/shiny/render.ts

```diff
@@ -354,6 +367,209 @@ function addStylesheetsAndRestyle(links: HTMLLinkElement[]): void {
   });
 }
 
+function getStylesheetLinkTags(dep: HtmlDepNormalized): HTMLLinkElement[] {
+  // Convert stylesheet objs to links early, because if `restyle` is true, we'll
+  // pass them through to `addStylesheetsAndRestyle` below.
+  return dep.stylesheet.map((x) => {
+    // Add "rel" and "type" fields if not already present.
+    if (!hasDefinedProperty(x, "rel")) x.rel = "stylesheet";
+    if (!hasDefinedProperty(x, "type")) x.type = "text/css";
+
+    const link = document.createElement("link");
+
+    Object.entries(x).forEach(function ([attr, val]: [
+      string,
+      string | undefined
+    ]) {
+      if (attr === "href") {
+        val = encodeURI(val as string);
+      }
+      // If val isn't truthy (e.g., null), consider it a boolean attribute
+      link.setAttribute(attr, val ? val : "");
+    });
+
+    return link;
+  });
+}
+
+function appendStylesheetLinkTags(
+  dep: HtmlDepNormalized,
+  $head: JQuery<HTMLElement>
+): void {
+  const stylesheetLinks = getStylesheetLinkTags(dep);
+
+  if (stylesheetLinks.length !== 0) {
+    $head.append(stylesheetLinks);
+  }
+}
+
+function appendScriptTags(dep: HtmlDepNormalized, $head: JQuery<HTMLElement>) {
+  dep.script.forEach((x) => {
+    const script = document.createElement("script");
+
+    Object.entries(x).forEach(function ([attr, val]) {
+      if (attr === "src") {
+        val = encodeURI(val);
+      }
+      // If val isn't truthy (e.g., null), consider it a boolean attribute
+      script.setAttribute(attr, val ? val : "");
+    });
+
+    $head.append(script);
+  });
+}
+
+async function appendScriptTagsAsync(dep: HtmlDepNormalized): Promise<void> {
+  const scriptPromises: Array<Promise<any>> = [];
+  const scriptElements: HTMLScriptElement[] = [];
+
+  dep.script.forEach((x) => {
+    const script = document.createElement("script");
+
+    if (!hasDefinedProperty(x, "async")) {
+      // Set async to false by default, so that if there are multiple script
+      // tags, they are guaranteed to run in order. For dynamically added
+      // <script> tags, browsers set async to true by default, which differs
+      // from static <script> tags in the html, which default to false.
+      //
+      // Refs:
+      // https://stackoverflow.com/a/8996894/412655
+      // https://jason-ge.medium.com/dynamically-load-javascript-files-in-order-5318ac6bcc61
+      //
+      // Note that one odd thing about these dynamically-created <script> tags
+      // is that even though the JS object's `x.script` property is true, it
+      // does NOT show up as a property on the <script> element.
+      script.async = false;
+    }
+
+    Object.entries(x).forEach(function ([attr, val]) {
+      if (attr === "src") {
+        val = encodeURI(val);
```

Comment:
It'll get over-escaped. This escaping was already present, so it seems the expectation has been that it will receive an unescaped URL from the server side.

```js
encodeURI("%")
# '%25'
```